File: rdp.go
package main

import (
	"bufio"
	"bytes"
	"crypto/x509"
	"encoding/asn1"
	"encoding/base64"
	"errors"
	"fmt"
	"math/big"
	"os"
	"strings"
	"time"
	"unicode/utf16"
)

// readRdpText читает .rdp с учётом кодировки. mstsc пишет UTF-16LE (BOM FF FE),
// но встречаются UTF-8 (в т.ч. с BOM), а изредка — UTF-16LE без BOM (после
// правки сторонним редактором). Ключи всегда ASCII.
func readRdpText(path string) (string, error) {
	b, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}
	switch {
	case len(b) >= 2 && b[0] == 0xFF && b[1] == 0xFE: // UTF-16LE BOM
		return decodeUTF16(b[2:], true), nil
	case len(b) >= 2 && b[0] == 0xFE && b[1] == 0xFF: // UTF-16BE BOM
		return decodeUTF16(b[2:], false), nil
	case len(b) >= 3 && b[0] == 0xEF && b[1] == 0xBB && b[2] == 0xBF: // UTF-8 BOM
		return string(b[3:]), nil
	case len(b) >= 2 && b[1] == 0x00 && b[0] != 0x00: // UTF-16LE без BOM (ASCII-старт)
		return decodeUTF16(b, true), nil
	}
	return string(b), nil // UTF-8 / ANSI
}

func decodeUTF16(b []byte, le bool) string {
	u := make([]uint16, 0, len(b)/2)
	for i := 0; i+1 < len(b); i += 2 {
		if le {
			u = append(u, uint16(b[i])|uint16(b[i+1])<<8)
		} else {
			u = append(u, uint16(b[i])<<8|uint16(b[i+1]))
		}
	}
	return string(utf16.Decode(u))
}

// rdpValue возвращает значение поля key (часть после «key:тип:»). Строки .rdp
// имеют вид «имя:тип:значение», сравнение имени без учёта регистра.
func rdpValue(text, key string) (string, bool) {
	sc := bufio.NewScanner(strings.NewReader(text))
	sc.Buffer(make([]byte, 0, 64*1024), 16*1024*1024) // подпись — одна длинная строка
	kp := strings.ToLower(key) + ":"
	for sc.Scan() {
		line := strings.TrimSpace(sc.Text())
		if strings.HasPrefix(strings.ToLower(line), kp) {
			if parts := strings.SplitN(line, ":", 3); len(parts) == 3 {
				return parts[2], true
			}
		}
	}
	return "", false
}

// isRemoteAppRdp: файл описывает сеанс RemoteApp (remoteapplicationmode:i:1).
func isRemoteAppRdp(text string) bool {
	v, ok := rdpValue(text, "remoteapplicationmode")
	return ok && strings.TrimSpace(v) == "1"
}

var oidSignedData = asn1.ObjectIdentifier{1, 2, 840, 113549, 1, 7, 2}

type pkcsContentInfo struct {
	ContentType asn1.ObjectIdentifier
	Content     asn1.RawValue `asn1:"explicit,optional,tag:0"`
}

// pkcsSignedData — SignedData ровно до signerInfos: сертификаты + первый
// SignerInfo (по его issuerAndSerial выбираем именно подписавший сертификат).
type pkcsSignedData struct {
	Version      int
	DigestAlgos  asn1.RawValue `asn1:"set"`
	ContentInfo  asn1.RawValue
	Certificates asn1.RawValue `asn1:"optional,tag:0"`
	CRLs         asn1.RawValue `asn1:"optional,tag:1"`
	SignerInfos  asn1.RawValue `asn1:"set"`
}

type issuerAndSerial struct {
	Issuer asn1.RawValue
	Serial *big.Int
}

// signerInfoHead — начало SignerInfo (version + issuerAndSerial); остаток
// (digestAlgorithm и далее) asn1.Unmarshal вернёт как rest.
type signerInfoHead struct {
	Version int
	IAS     issuerAndSerial
}

func stripSpace(s string) string {
	return strings.Map(func(r rune) rune {
		switch r {
		case ' ', '\t', '\r', '\n':
			return -1
		}
		return r
	}, s)
}

// parseCerts разбирает конкатенацию DER-сертификатов по одному, пропуская те,
// что x509 не осилил (иначе один странный сертификат в мешке скрыл бы дату).
func parseCerts(der []byte) []*x509.Certificate {
	var out []*x509.Certificate
	for len(der) > 0 {
		var raw asn1.RawValue
		rest, err := asn1.Unmarshal(der, &raw)
		if err != nil {
			break
		}
		if c, err := x509.ParseCertificate(raw.FullBytes); err == nil {
			out = append(out, c)
		}
		der = rest
	}
	return out
}

// signerCertExpiry: NotAfter именно подписавшего сертификата. Подписанта берём
// из первого SignerInfo (issuer+serial); если не удалось сопоставить — фолбэк
// на не-CA сертификат с самой ранней датой (а если все CA — самый ранний).
func signerCertExpiry(certs []*x509.Certificate, signerInfos []byte) (time.Time, bool) {
	var si signerInfoHead
	if _, err := asn1.Unmarshal(signerInfos, &si); err == nil && si.IAS.Serial != nil {
		for _, c := range certs {
			if c.SerialNumber != nil && c.SerialNumber.Cmp(si.IAS.Serial) == 0 &&
				bytes.Equal(c.RawIssuer, si.IAS.Issuer.FullBytes) {
				return c.NotAfter, true
			}
		}
	}
	var best time.Time
	found := false
	for _, c := range certs {
		if c.IsCA {
			continue
		}
		if !found || c.NotAfter.Before(best) {
			best, found = c.NotAfter, true
		}
	}
	if !found {
		for _, c := range certs {
			if !found || c.NotAfter.Before(best) {
				best, found = c.NotAfter, true
			}
		}
	}
	return best, found
}

// signatureExpiry: значение поля signature:s: — base64 с заголовком rdpsign
// (12 байт) перед PKCS#7. base64 в .rdp разбит пробелами каждые 64 символа.
// Заголовок пропускаем, находя ContentInfo по OID signedData; из PKCS#7 берём
// сертификаты и дату окончания подписавшего.
func signatureExpiry(sigB64 string) (time.Time, bool, error) {
	blob, err := base64.StdEncoding.DecodeString(stripSpace(sigB64))
	if err != nil {
		return time.Time{}, false, err
	}
	oid := []byte{0x06, 0x09, 0x2A, 0x86, 0x48, 0x86, 0xF7, 0x0D, 0x01, 0x07, 0x02}
	idx := bytes.Index(blob, oid)
	if idx < 0 {
		return time.Time{}, false, errors.New("OID signedData не найден")
	}
	var ci pkcsContentInfo
	parsed := false
	for hdr := 2; hdr <= 7; hdr++ { // размер заголовка SEQUENCE перед OID (30 [8x LL...])
		start := idx - hdr
		if start < 0 {
			continue
		}
		if _, e := asn1.Unmarshal(blob[start:], &ci); e == nil && ci.ContentType.Equal(oidSignedData) {
			parsed = true
			break
		}
	}
	if !parsed {
		return time.Time{}, false, errors.New("не удалось разобрать ContentInfo")
	}
	var sd pkcsSignedData
	if _, err := asn1.Unmarshal(ci.Content.Bytes, &sd); err != nil {
		return time.Time{}, false, err
	}
	certs := parseCerts(sd.Certificates.Bytes)
	if len(certs) == 0 {
		return time.Time{}, false, errors.New("нет сертификатов в подписи")
	}
	exp, ok := signerCertExpiry(certs, sd.SignerInfos.Bytes)
	return exp, ok, nil
}

// expiryFromText: срок действия подписи по уже прочитанному тексту .rdp (файл
// читаем один раз на пути «проверили RemoteApp → показали дату»). ok=false —
// если файл не подписан или подпись не удалось разобрать (дату не показываем).
func expiryFromText(text string) (time.Time, bool) {
	sig, ok := rdpValue(text, "signature")
	if !ok || strings.TrimSpace(sig) == "" {
		return time.Time{}, false
	}
	exp, ok, err := signatureExpiry(sig)
	if err != nil {
		return time.Time{}, false
	}
	return exp, ok
}

// formatDate: дата в виде ДД.ММ.ГГГГ (в UTC, как хранится в сертификате).
func formatDate(t time.Time) string {
	u := t.UTC()
	return fmt.Sprintf("%02d.%02d.%04d", u.Day(), int(u.Month()), u.Year())
}

File: rdp.c
// Чтение .rdp с учётом кодировки; извлечение полей; дата окончания подписи (.rdp
// подписан PKCS#7 в поле signature:s:). Разбор подписи — через CryptoAPI.
#include "rdpkey.h"
#include <wincrypt.h>
#include <stdlib.h>
#include <string.h>
#include <wchar.h>
#include <stdio.h>

static BOOL read_bytes(const wchar_t *path, unsigned char **out, int *outn) {
    HANDLE h = CreateFileW(path, GENERIC_READ, FILE_SHARE_READ, NULL,
                           OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
    if (h == INVALID_HANDLE_VALUE) return FALSE;
    DWORD sz = GetFileSize(h, NULL);
    unsigned char *buf = (unsigned char *)malloc(sz + 1);
    if (!buf) { CloseHandle(h); return FALSE; }
    DWORD rd = 0;
    BOOL ok = ReadFile(h, buf, sz, &rd, NULL);
    CloseHandle(h);
    if (!ok) { free(buf); return FALSE; }
    buf[rd] = 0;
    *out = buf; *outn = (int)rd;
    return TRUE;
}

static wchar_t *decode_utf16(const unsigned char *b, int n, BOOL le) {
    int cnt = n / 2;
    wchar_t *w = (wchar_t *)malloc((cnt + 1) * sizeof(wchar_t));
    if (!w) return NULL;
    for (int i = 0; i < cnt; i++) {
        unsigned c = le ? (b[2*i] | (b[2*i+1] << 8)) : ((b[2*i] << 8) | b[2*i+1]);
        w[i] = (wchar_t)c;
    }
    w[cnt] = 0;
    return w;
}

static wchar_t *decode_utf8(const unsigned char *b, int n) {
    int wn = MultiByteToWideChar(CP_UTF8, 0, (const char *)b, n, NULL, 0);
    wchar_t *w = (wchar_t *)malloc((wn + 1) * sizeof(wchar_t));
    if (!w) return NULL;
    MultiByteToWideChar(CP_UTF8, 0, (const char *)b, n, w, wn);
    w[wn] = 0;
    return w;
}

BOOL read_rdp_text(const wchar_t *path, wchar_t **outText) {
    unsigned char *b; int n;
    if (!read_bytes(path, &b, &n)) return FALSE;
    wchar_t *w = NULL;
    if (n >= 2 && b[0] == 0xFF && b[1] == 0xFE)                 w = decode_utf16(b + 2, n - 2, TRUE);
    else if (n >= 2 && b[0] == 0xFE && b[1] == 0xFF)           w = decode_utf16(b + 2, n - 2, FALSE);
    else if (n >= 3 && b[0] == 0xEF && b[1] == 0xBB && b[2] == 0xBF) w = decode_utf8(b + 3, n - 3);
    else if (n >= 2 && b[1] == 0x00 && b[0] != 0x00)          w = decode_utf16(b, n, TRUE);
    else                                                       w = decode_utf8(b, n);
    free(b);
    if (!w) return FALSE;
    *outText = w;
    return TRUE;
}

// find_value_range: для строки «key:тип:значение» вернуть [vs,ve) значения (после
// второго ':'), без копирования (подпись — одна очень длинная строка).
static BOOL find_value_range(const wchar_t *text, const wchar_t *keyColon,
                             const wchar_t **vs, const wchar_t **ve) {
    int kl = (int)wcslen(keyColon);
    const wchar_t *p = text;
    while (*p) {
        const wchar_t *nl = wcschr(p, L'\n');
        const wchar_t *lineEnd = nl ? nl : p + wcslen(p);
        const wchar_t *q = p;
        while (q < lineEnd && (*q == L' ' || *q == L'\t')) q++;
        if ((lineEnd - q) >= kl && _wcsnicmp(q, keyColon, kl) == 0) {
            const wchar_t *t = q + kl;               // тип
            const wchar_t *colon = t;
            while (colon < lineEnd && *colon != L':') colon++;
            if (colon < lineEnd) {
                const wchar_t *v = colon + 1;
                const wchar_t *e = lineEnd;
                while (e > v && (e[-1] == L'\r' || e[-1] == L' ' || e[-1] == L'\t')) e--;
                *vs = v; *ve = e;
                return TRUE;
            }
        }
        if (!nl) break;
        p = nl + 1;
    }
    return FALSE;
}

BOOL rdp_value(const wchar_t *text, const wchar_t *key, wchar_t *out, int outcch) {
    wchar_t kc[128];
    swprintf(kc, 128, L"%ls:", key);
    const wchar_t *vs, *ve;
    if (!find_value_range(text, kc, &vs, &ve)) return FALSE;
    // ведущие пробелы значения тоже уберём (Go сравнивает TrimSpace)
    while (vs < ve && (*vs == L' ' || *vs == L'\t')) vs++;
    int len = (int)(ve - vs);
    if (len >= outcch) len = outcch - 1;
    wcsncpy(out, vs, len);
    out[len] = 0;
    return TRUE;
}

BOOL is_remoteapp_rdp(const wchar_t *text) {
    wchar_t v[16];
    if (!rdp_value(text, L"remoteapplicationmode", v, 16)) return FALSE;
    return wcscmp(v, L"1") == 0;
}

#ifdef NO_CERT
BOOL expiry_from_text(const wchar_t *text, FILETIME *out) { (void)text; (void)out; return FALSE; }
#else
// signature_expiry: base64 -> blob (12-байт заголовок rdpsign + PKCS#7). Находим
// ContentInfo по OID signedData, грузим PKCS#7 через CryptQueryObject и берём
// NotAfter именно подписавшего сертификата (CMSG_SIGNER_CERT_INFO_PARAM).
static BOOL signature_expiry(const char *b64, FILETIME *out) {
    DWORD blen = 0;
    if (!CryptStringToBinaryA(b64, 0, CRYPT_STRING_BASE64, NULL, &blen, NULL, NULL) || blen == 0)
        return FALSE;
    BYTE *blob = (BYTE *)malloc(blen);
    if (!blob) return FALSE;
    if (!CryptStringToBinaryA(b64, 0, CRYPT_STRING_BASE64, blob, &blen, NULL, NULL)) {
        free(blob);
        return FALSE;
    }

    static const BYTE oid[] = {0x06,0x09,0x2A,0x86,0x48,0x86,0xF7,0x0D,0x01,0x07,0x02};
    int idx = -1;
    for (DWORD i = 0; i + sizeof(oid) <= blen; i++) {
        if (memcmp(blob + i, oid, sizeof(oid)) == 0) { idx = (int)i; break; }
    }
    if (idx < 0) { free(blob); return FALSE; }

    BOOL ok = FALSE;
    for (int hdr = 2; hdr <= 7 && !ok; hdr++) {
        int start = idx - hdr;
        if (start < 0 || blob[start] != 0x30) continue;
        CERT_BLOB cb;
        cb.cbData = blen - start;
        cb.pbData = blob + start;
        HCERTSTORE hStore = NULL;
        HCRYPTMSG hMsg = NULL;
        if (CryptQueryObject(CERT_QUERY_OBJECT_BLOB, &cb,
                CERT_QUERY_CONTENT_FLAG_PKCS7_SIGNED, CERT_QUERY_FORMAT_FLAG_BINARY,
                0, NULL, NULL, NULL, &hStore, &hMsg, NULL)) {
            DWORD need = 0;
            if (CryptMsgGetParam(hMsg, CMSG_SIGNER_CERT_INFO_PARAM, 0, NULL, &need) && need > 0) {
                CERT_INFO *ci = (CERT_INFO *)malloc(need);
                if (ci && CryptMsgGetParam(hMsg, CMSG_SIGNER_CERT_INFO_PARAM, 0, ci, &need)) {
                    PCCERT_CONTEXT signer = CertGetSubjectCertificateFromStore(
                        hStore, X509_ASN_ENCODING | PKCS_7_ASN_ENCODING, ci);
                    if (signer) {
                        *out = signer->pCertInfo->NotAfter;
                        ok = TRUE;
                        CertFreeCertificateContext(signer);
                    }
                }
                free(ci);
            }
            if (hMsg) CryptMsgClose(hMsg);
            if (hStore) CertCloseStore(hStore, 0);
        }
    }
    free(blob);
    return ok;
}

BOOL expiry_from_text(const wchar_t *text, FILETIME *out) {
    const wchar_t *vs, *ve;
    if (!find_value_range(text, L"signature:", &vs, &ve)) return FALSE;
    int cap = (int)(ve - vs);
    if (cap <= 0) return FALSE;
    char *b64 = (char *)malloc(cap + 1);
    if (!b64) return FALSE;
    int k = 0;
    for (const wchar_t *p = vs; p < ve; p++) {
        wchar_t c = *p;
        if (c == L' ' || c == L'\t' || c == L'\r' || c == L'\n') continue;
        if (c < 128) b64[k++] = (char)c;
    }
    b64[k] = 0;
    BOOL ok = (k > 0) && signature_expiry(b64, out);
    free(b64);
    return ok;
}
#endif // NO_CERT
